torch/distributed/pipelining/schedules.py

```diff
@@ -2097,71 +2166,331 @@ def backward_stage_index(step):
+    ):
+        self.pp_group_size = stages[0].group_size
+        super().__init__(
+            stages=stages,
+            n_microbatches=n_microbatches,
+            loss_fn=loss_fn,
+            args_chunk_spec=args_chunk_spec,
+            kwargs_chunk_spec=kwargs_chunk_spec,
+            output_merge_spec=output_merge_spec,
+            use_full_backward=False,
+        )
+        self.n_local_stages = len(stages)
+        self.rank = stages[0].group_rank
+        if self.n_local_stages % 2 != 0:
+            raise ValueError(
+                "ZBV schedule requires the number of local chunks to be "
+                f"divisible by 2, but got {len(stages)}."
+            )
 
-            temp_seen_ops: Set[Tuple[int, _ComputationType, int]] = set()
+        # 1. Create the pipeline_order (all ranks do this calculation)
+        # This will be used to keep track of the current state of the entire pipeline
+        # pipeline_order[rank] = [Action(computation_type, microbatch_index, stage_index), ...]
+        self.pipeline_order: Dict[int, List[Optional[_Action]]] = self._get_zbv_schedule(
+            len(stages),
+            self.pp_group_size,
+            n_microbatches,
+        )
+        self.pipeline_order = _add_bubbles_to_actions(
+            self.n_local_stages * self.pp_group_size,
+            self.pipeline_order,
+            self.pp_group_size,
+        )
 
-            for rank in range(self.pp_group_size):
-                timestamp = next_pointer[rank]
-                if timestamp >= len(actions[rank]):
+    def _get_zbv_schedule(
+        self,
+        num_model_chunks,
+        pipeline_parallel_size,
+        num_microbatches,
+    ):
+        pp_group_size = pipeline_parallel_size
+        mb_size = num_microbatches
+        num_local_stages = num_model_chunks
+        compute_schedules = {}
+        category_map = {
+            F: 0,
+            I: 1,
+            W: 2,
+        }
+        chunk_0 = 0
+        chunk_1 = 1
+        def get_compute_schedule(pipeline_parallel_size, num_microbatches):
+            num_chunks = 2
+            n_node = len(category_map) * num_chunks * pipeline_parallel_size * num_microbatches
+
+            def get_id(op, chunk, rank, microbatch_id):
+                category = category_map[op]
+                return (
+                    category * 2 * pipeline_parallel_size * num_microbatches
+                    + chunk * pipeline_parallel_size * num_microbatches
+                    + rank * num_microbatches
+                    + microbatch_id
+                )
+            ops_count = []
+
+            for i in range(pipeline_parallel_size):
+                temp = {}
+                temp[F] = [0] * 2
+                temp[I] = [0] * 2
+                temp[W] = [0] * 2
+                ops_count.append(temp)
+
+            fbw_mem = [3, -1, -2]
+            max_mem = 3 * (pipeline_parallel_size * 2)
+            end_time = [-1] * n_node
```

Comment:
as i read through the code below, i realize i'm confused between cur_time and end_time.  Maybe it will become more clear..